README
This is nss_nonlocal, an nsswitch module that acts as a proxy for other 
nsswitch modules like hesiod, but prevents non-local users from 
potentially gaining local privileges by spoofing local UIDs and GIDs.

To use it, configure /etc/nsswitch.conf as follows:

passwd:         compat nonlocal
passwd_nonlocal: hesiod
group:          compat nonlocal
group_nonlocal: hesiod

The module also assigns special properties to two local groups, if
they exist:

• If the local group ‘nss-nonlocal-users’ exists, then nonlocal users
  will be automatically added to it.

• If the local group ‘nss-local-users’ exists, then local users will
  be automatically added to it.

Copyright © 2007–2010 Anders Kaseorg <andersk@mit.edu> and Tim Abbott
<tabbott@mit.edu>

nss_nonlocal is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation; either version 2.1 of the
License, or (at your option) any later version.

nss_nonlocal is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with nss_nonlocal; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301  USA


debian/changelog
libnss-nonlocal (2.2-0debathena1) UNRELEASED; urgency=medium

  * New upstream release.
    - Support glibc 2.27.

 -- Anders Kaseorg <andersk@mit.edu>  Sat, 05 May 2018 14:15:17 -0400

libnss-nonlocal (2.1-0debathena1) unstable; urgency=low

  * New upstream release.
    - Support Automake 1.12.
    - Guard one-time initialization with memory barriers.
    - Make initgroups_dyn succeed when adding only magic groups.
  * Rewrite packaging with Debhelper 7.
  * Move magic user and group creation to separate libnss-nonlocal-common
    package.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 13 Jun 2013 04:43:16 -0400

libnss-nonlocal (2.0-0debathena2) unstable; urgency=low

  * Pre-Depend adduser.
  * Don’t Build-Depend debhelper 9.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 16 Aug 2012 20:38:35 -0400

libnss-nonlocal (2.0-0debathena1) unstable; urgency=low

  * New upstream release.
    - Fix errno saving and restoring.
    - Document nss-nonlocal-users and nss-local-users groups in README.
    - Allow local whitelisting of nonlocal user and group memberships,
      using the magic local ‘nss-nonlocal-users’ user and group.
  * Create magic ‘nss-nonlocal-users’ user.
  * Remove obsolete exclusion of glibc-private dependency.
  * Use automake 1.10 and autoconf 2.61.
  * Add multiarch support for distros from the future.
  * Disable dh_buildinfo, which currently fails at multiarch.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 30 Mar 2011 04:57:36 -0400

libnss-nonlocal (1.11-0debathena3) unstable; urgency=low

  * Remove the workaround for buggy CDBS $(CC) splitting (fixed upstream).
  * Fix cross-compiling on karmic and lucid.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 23 Jun 2010 21:17:50 -0400

libnss-nonlocal (1.11-0debathena2) unstable; urgency=low

  * Install lib32nss-nonlocal to the right place on Ubuntu.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 23 Jun 2010 03:50:33 -0400

libnss-nonlocal (1.11-0debathena1) unstable; urgency=low

  * New upstream release.
    - Check that a nonlocal lookup by id returns the right id.
    - Update licensing to LGPLv2.1+.
    - Change syslog priority when removing local groups from non-local
      users
  * Update build flags for biarch packages (fixes lib32nss-nonlocal path
    on squeeze).
  * Fix FTBFS on squeeze due to CDBS’s buggy splitting of multiword
    $(CC) (Debian bug #576967).
  * Disable CDBS cross-building logic, to fix a mysterious FTBFS where
    configure --host fails to find nm.
  * Bump Standards-Version to 3.8.4 (no changes required).
  * Bump Debhelper compatibility level to 6.
  * Set maintainer to me.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 03 May 2010 00:45:14 -0400

libnss-nonlocal (1.10-0debathena1) unstable; urgency=low

  * New upstream version.
    - Disallow numeric nonlocal user/group names that look like local
      uid/gids.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 23 Oct 2009 13:49:43 -0400

libnss-nonlocal (1.9-0debathena2) unstable; urgency=low

  * Also build a biarch lib32nss-nonlocal or lib64nss-nonlocal package.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 19 Jun 2009 23:54:38 -0400

libnss-nonlocal (1.9-0debathena1) unstable; urgency=low

  * New upstream version.
    - Corrects an out-of-memory error in the presence of very large local
      groups.

 -- Anders Kaseorg <andersk@mit.edu>  Sun, 24 May 2009 17:30:08 -0400

libnss-nonlocal (1.8-0debathena4) unstable; urgency=low

  * Only exclude glibc-private for new enough libc6, because old
    dpkg-shlibdeps does not support -x.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 05 Feb 2009 01:59:08 -0500

libnss-nonlocal (1.8-0debathena3) unstable; urgency=low

  * Force exclusion of glibc-private dependency.  This is necessary for
    libnss-nonlocal to build and install with libc6 2.9 in Ubuntu Jaunty
    and Debian experimental, which now generates an invalid dependency on
    glibc-private for uses of private glibc symbols.

 -- Anders Kaseorg <andersk@mit.edu>  Sun, 01 Feb 2009 00:26:31 -0500

libnss-nonlocal (1.8-0debathena2) unstable; urgency=low

  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.
  * Clean up debian/copyright.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 18:03:22 -0500

libnss-nonlocal (1.8-0debathena1) unstable; urgency=low

  * Correct the buffer size passed to realloc.

 -- Anders Kaseorg <andersk@mit.edu>  Sat, 09 Aug 2008 22:41:57 -0400

libnss-nonlocal (1.7-0debathena1) unstable; urgency=low

  * Replace the magic buflen hack with explicit iteration over the nss
    chain, so that getpwent/getgrent works with nscd enabled.
  * Fix some memory leaks.
  * Autotoolfiscate.
  * Code cleanup.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 29 Jul 2008 06:57:53 -0400

libnss-nonlocal (1.6-0debathena1) unstable; urgency=low

  * Use a version script to hide internal symbols.

 -- Anders Kaseorg <andersk@mit.edu>  Sat, 15 Mar 2008 06:42:02 -0400

libnss-nonlocal (1.5-0debathena1) unstable; urgency=low

  * Add support for NSS_NONLOCAL_IGNORE environment variable, which
    causes nss_nonlocal to pretend there are no nonlocal users.  This
    allows us to support letting adduser add users and groups with the
    same name as a nonlocal user/group.

 -- Tim Abbott <tabbott@mit.edu>  Mon, 25 Feb 2008 19:12:30 -0500

libnss-nonlocal (1.4-0debathena1) unstable; urgency=low


nonlocal.h
/*
 * nonlocal.h
 * common definitions for nss_nonlocal proxy
 *
 * Copyright © 2007–2010 Anders Kaseorg <andersk@mit.edu> and Tim
 * Abbott <tabbott@mit.edu>
 *
 * This file is part of nss_nonlocal.
 *
 * nss_nonlocal is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * nss_nonlocal is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with nss_nonlocal; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

#ifndef NONLOCAL_H
#define NONLOCAL_H

#include "config.h"
#include "nsswitch-internal.h"

struct walk_nss {
    enum nss_status *status;
    int (*lookup)(service_user **ni, const char *fct_name,
		  void **fctp) internal_function;
    const char *fct_name;
    int *errnop;
    char **buf;
    size_t *buflen;
};

enum nss_status check_nonlocal_uid(const char *user, uid_t uid, int *errnop);
enum nss_status check_nonlocal_gid(const char *user, gid_t gid, int *errnop);
enum nss_status check_nonlocal_user(const char *user, int *errnop);

#define NONLOCAL_IGNORE_ENV "NSS_NONLOCAL_IGNORE"

#endif /* NON_LOCAL_H */


walk_nss.h
/*
 * walk_nss.h
 * NSS walking template for nss_nonlocal proxy
 *
 * Copyright © 2011 Anders Kaseorg <andersk@mit.edu> and Tim Abbott
 * <tabbott@mit.edu>
 *
 * This file is part of nss_nonlocal.
 *
 * nss_nonlocal is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * nss_nonlocal is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with nss_nonlocal; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

{
    static bool initialized = false;
    static service_user *startp;
    static void *fct_start;

    service_user *nip;
    union {
	__typeof__(self) l;
	void *ptr;
    } fct;
    int old_errno = errno;

    if (!initialized) {
	if (w.lookup(&startp, w.fct_name, &fct_start) != 0) {
	    *w.status = NSS_STATUS_UNAVAIL;
	    goto walk_nss_out;
	}
	__sync_synchronize();
	initialized = true;
    }

    nip = startp;
    fct.ptr = fct_start;

    if (w.buf != NULL) {
	*w.buf = malloc(*w.buflen);
	errno = old_errno;
	if (*w.buf == NULL) {
	    *w.status = NSS_STATUS_TRYAGAIN;
	    *w.errnop = ENOMEM;
	    goto walk_nss_out;
	}
    }

    do {
    walk_nss_morebuf:
	if (fct.ptr == NULL)
	    *w.status = NSS_STATUS_UNAVAIL;
	else if (self != NULL && fct.l == self)
	    *w.status = NSS_STATUS_NOTFOUND;
	else
	    *w.status = DL_CALL_FCT(fct.l, args);
	if (*w.status == NSS_STATUS_TRYAGAIN &&
	    w.errnop != NULL && *w.errnop == ERANGE) {
	    if (w.buf == NULL)
		break;
	    free(*w.buf);
	    *w.buflen *= 2;
	    *w.buf = malloc(*w.buflen);
	    errno = old_errno;
	    if (*w.buf == NULL) {
		*w.errnop = ENOMEM;
		goto walk_nss_out;
	    }
	    goto walk_nss_morebuf;
	}
    } while (__nss_next(&nip, w.fct_name, &fct.ptr, *w.status, 0) == 0);

    if (w.buf != NULL && *w.status != NSS_STATUS_SUCCESS) {
	free(*w.buf);
	*w.buf = NULL;
    }

 walk_nss_out:
    ;
}


nonlocal-shadow.c
/*
 * nonlocal-shadow.c
 * shadow database for nss_nonlocal proxy.
 *
 * Copyright © 2007–2010 Anders Kaseorg <andersk@mit.edu>
 *
 * This file is part of nss_nonlocal.
 *
 * nss_nonlocal is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * nss_nonlocal is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with nss_nonlocal; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

#define _GNU_SOURCE

#include <sys/types.h>
#include <dlfcn.h>
#include <errno.h>
#include <nss.h>
#include <shadow.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>

#include "nsswitch-internal.h"
#include "nonlocal.h"


static service_user *__nss_shadow_nonlocal_database;

static int
internal_function
__nss_shadow_nonlocal_lookup(service_user **ni, const char *fct_name,
			    void **fctp)
{
    if (__nss_shadow_nonlocal_database == NULL
	&& __nss_database_lookup("shadow_nonlocal", NULL, NULL,
				 &__nss_shadow_nonlocal_database) < 0)
	return -1;

    *ni = __nss_shadow_nonlocal_database;

    *fctp = __nss_lookup_function(*ni, fct_name);
    return 0;
}


static service_user *spent_startp, *spent_nip;
static void *spent_fct_start;
static union {
    enum nss_status (*l)(struct spwd *pwd, char *buffer, size_t buflen,
			 int *errnop);
    void *ptr;
} spent_fct;
static const char *spent_fct_name = "getspent_r";

enum nss_status
_nss_nonlocal_setspent(int stayopen)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = &__nss_shadow_nonlocal_lookup, .fct_name = "setspent",
	.status = &status
    };
    const __typeof__(&_nss_nonlocal_setspent) self = NULL;
#define args (stayopen)
#include "walk_nss.h"
#undef args
    if (status != NSS_STATUS_SUCCESS)
	return status;

    if (spent_fct_start == NULL)
	__nss_shadow_nonlocal_lookup(&spent_startp, spent_fct_name,
				     &spent_fct_start);
    spent_nip = spent_startp;
    spent_fct.ptr = spent_fct_start;
    return NSS_STATUS_SUCCESS;
}

enum nss_status
_nss_nonlocal_endspent(void)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = &__nss_shadow_nonlocal_lookup, .fct_name = "endspent",
	.status = &status
    };
    const __typeof__(&_nss_nonlocal_endspent) self = NULL;

    spent_nip = NULL;

#define args ()
#include "walk_nss.h"
#undef args
    return status;
}

enum nss_status
_nss_nonlocal_getspent_r(struct spwd *pwd, char *buffer, size_t buflen,
			 int *errnop)
{
    enum nss_status status;

    char *nonlocal_ignore = getenv(NONLOCAL_IGNORE_ENV);
    if (nonlocal_ignore != NULL && nonlocal_ignore[0] != '\0')
	return NSS_STATUS_UNAVAIL;

    if (spent_nip == NULL) {
	status = _nss_nonlocal_setspent(0);
	if (status != NSS_STATUS_SUCCESS)
	    return status;
    }
    do {
	if (spent_fct.ptr == NULL)
	    status = NSS_STATUS_UNAVAIL;
	else
	    status = DL_CALL_FCT(spent_fct.l, (pwd, buffer, buflen, errnop));	
	if (status == NSS_STATUS_TRYAGAIN && *errnop == ERANGE)
	    return status;

	if (status == NSS_STATUS_SUCCESS)
	    return NSS_STATUS_SUCCESS;
    } while (__nss_next(&spent_nip, spent_fct_name, &spent_fct.ptr, status, 0) == 0);

    spent_nip = NULL;
    return NSS_STATUS_NOTFOUND;
}


enum nss_status
_nss_nonlocal_getspnam_r(const char *name, struct spwd *pwd,
			 char *buffer, size_t buflen, int *errnop)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = __nss_shadow_nonlocal_lookup, .fct_name = "getspnam_r",
	.status = &status, .errnop = errnop
    };
    const __typeof__(&_nss_nonlocal_getspnam_r) self = NULL;
#define args (name, pwd, buffer, buflen, errnop)
#include "walk_nss.h"
#undef args
    if (status != NSS_STATUS_SUCCESS)
	return status;

    if (strcmp(name, pwd->sp_namp) != 0) {
	syslog(LOG_ERR, "nss_nonlocal: discarding shadow %s from lookup for shadow %s\n", pwd->sp_namp, name);
	return NSS_STATUS_NOTFOUND;
    }

    return NSS_STATUS_SUCCESS;
}


nonlocal-group.c
/*
 * nonlocal-group.c
 * group database for nss_nonlocal proxy
 *
 * Copyright © 2007–2010 Anders Kaseorg <andersk@mit.edu> and Tim
 * Abbott <tabbott@mit.edu>
 *
 * This file is part of nss_nonlocal.
 *
 * nss_nonlocal is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * nss_nonlocal is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with nss_nonlocal; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

#define _GNU_SOURCE

#include <sys/types.h>
#include <dlfcn.h>
#include <errno.h>
#include <grp.h>
#include <nss.h>
#include <pwd.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

#include "nsswitch-internal.h"
#include "nonlocal.h"

/*
 * If the MAGIC_NONLOCAL_GROUPNAME local group exists, then nonlocal
 * users will be automatically added to it.  Furthermore, if a local
 * user is added to this group, then that user will inherit any
 * nonlocal gids from a nonlocal user of the same name, as
 * supplementary gids.
 */
#define MAGIC_NONLOCAL_GROUPNAME "nss-nonlocal-users"

/*
 * If the MAGIC_LOCAL_GROUPNAME local group exists, then local users
 * will be automatically added to it.
 */
#define MAGIC_LOCAL_GROUPNAME "nss-local-users"

/*
 * If the MAGIC_NONLOCAL_USERNAME local user is added to a local
 * group, then the local group will inherit the nonlocal membership of
 * a group of the same gid.
 */
#define MAGIC_NONLOCAL_USERNAME "nss-nonlocal-users"


enum nss_status
_nss_nonlocal_getgrnam_r(const char *name, struct group *grp,
			 char *buffer, size_t buflen, int *errnop);

enum nss_status
_nss_nonlocal_getgrgid_r(gid_t gid, struct group *grp,
			 char *buffer, size_t buflen, int *errnop);


static service_user *__nss_group_nonlocal_database;

static int
internal_function
__nss_group_nonlocal_lookup(service_user **ni, const char *fct_name,
			    void **fctp)
{
    if (__nss_group_nonlocal_database == NULL
	&& __nss_database_lookup("group_nonlocal", NULL, NULL,
				 &__nss_group_nonlocal_database) < 0)
	return -1;

    *ni = __nss_group_nonlocal_database;

    *fctp = __nss_lookup_function(*ni, fct_name);
    return 0;
}


enum nss_status
check_nonlocal_gid(const char *user, const char *group, gid_t gid, int *errnop)
{
    enum nss_status status;
    struct group gbuf;
    char *buf;
    size_t buflen = sysconf(_SC_GETGR_R_SIZE_MAX);
    const struct walk_nss w = {
	.lookup = &__nss_group_lookup, .fct_name = "getgrgid_r",
	.status = &status, .errnop = errnop, .buf = &buf, .buflen = &buflen
    };
    const __typeof__(&_nss_nonlocal_getgrgid_r) self = &_nss_nonlocal_getgrgid_r;
#define args (gid, &gbuf, buf, buflen, errnop)
#include "walk_nss.h"
#undef args

    if (status == NSS_STATUS_TRYAGAIN)
	return status;
    else if (status != NSS_STATUS_SUCCESS)
	return NSS_STATUS_SUCCESS;

    if (group == NULL || strcmp(gbuf.gr_name, group) == 0) {
	char *const *mem;
	for (mem = gbuf.gr_mem; *mem != NULL; mem++)
	    if (strcmp(*mem, MAGIC_NONLOCAL_USERNAME) == 0) {
		status = check_nonlocal_user(*mem, errnop);
		if (status == NSS_STATUS_TRYAGAIN) {
		    free(buf);
		    return status;
		} else if (status == NSS_STATUS_NOTFOUND) {
		    free(buf);
		    return NSS_STATUS_SUCCESS;
		}
		break;
	    }
    }

    syslog(LOG_DEBUG, "nss_nonlocal: removing local group %u (%s) from non-local user %s\n", gbuf.gr_gid, gbuf.gr_name, user);
    free(buf);
    return NSS_STATUS_NOTFOUND;
}

enum nss_status
check_nonlocal_group(const char *user, struct group *grp, int *errnop)
{
    enum nss_status status = NSS_STATUS_SUCCESS;
    int old_errno = errno;
    char *end;
    unsigned long gid;

    errno = 0;
    gid = strtoul(grp->gr_name, &end, 10);
    if (errno == 0 && *end == '\0' && (gid_t)gid == gid) {
	errno = old_errno;
	status = check_nonlocal_gid(user, grp->gr_name, gid, errnop);
    } else
	errno = old_errno;
    if (status != NSS_STATUS_SUCCESS)
	return status;

    return check_nonlocal_gid(user, grp->gr_name, grp->gr_gid, errnop);
}

enum nss_status
get_local_group(const char *name, struct group *grp, char **buffer, int *errnop)
{
    enum nss_status status;
    size_t buflen = sysconf(_SC_GETGR_R_SIZE_MAX);
    const struct walk_nss w = {
	.lookup = &__nss_group_lookup, .fct_name = "getgrnam_r",
	.status = &status, .errnop = errnop, .buf = buffer, .buflen = &buflen
    };
    const __typeof__(&_nss_nonlocal_getgrnam_r) self = &_nss_nonlocal_getgrnam_r;
#define args (name, grp, *buffer, buflen, errnop)
#include "walk_nss.h"
#undef args
    return status;
}

static service_user *grent_startp, *grent_nip;
static void *grent_fct_start;
static union {
    enum nss_status (*l)(struct group *grp, char *buffer, size_t buflen,
			 int *errnop);
    void *ptr;
} grent_fct;
static const char *grent_fct_name = "getgrent_r";

enum nss_status
_nss_nonlocal_setgrent(int stayopen)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = &__nss_group_nonlocal_lookup, .fct_name = "setgrent",
	.status = &status
    };
    const __typeof__(&_nss_nonlocal_setgrent) self = NULL;
#define args (stayopen)
#include "walk_nss.h"
#undef args
    if (status != NSS_STATUS_SUCCESS)
	return status;

    if (grent_fct_start == NULL)
	__nss_group_nonlocal_lookup(&grent_startp, grent_fct_name,
				    &grent_fct_start);
    grent_nip = grent_startp;
    grent_fct.ptr = grent_fct_start;
    return NSS_STATUS_SUCCESS;
}

enum nss_status
_nss_nonlocal_endgrent(void)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = &__nss_group_nonlocal_lookup, .fct_name = "endgrent",
	.status = &status
    };
    const __typeof__(&_nss_nonlocal_endgrent) self = NULL;

    grent_nip = NULL;

#define args ()
#include "walk_nss.h"
#undef args
    return status;
}

enum nss_status
_nss_nonlocal_getgrent_r(struct group *grp, char *buffer, size_t buflen,
			 int *errnop)
{
    enum nss_status status;

    char *nonlocal_ignore = getenv(NONLOCAL_IGNORE_ENV);
    if (nonlocal_ignore != NULL && nonlocal_ignore[0] != '\0')
	return NSS_STATUS_UNAVAIL;

    if (grent_nip == NULL) {
	status = _nss_nonlocal_setgrent(0);
	if (status != NSS_STATUS_SUCCESS)
	    return status;
    }
    do {
	if (grent_fct.ptr == NULL)
	    status = NSS_STATUS_UNAVAIL;
	else {
	    int nonlocal_errno;
	    do
		status = DL_CALL_FCT(grent_fct.l, (grp, buffer, buflen, errnop));
	    while (status == NSS_STATUS_SUCCESS &&
		   check_nonlocal_group("(unknown)", grp, &nonlocal_errno) != NSS_STATUS_SUCCESS);
	}
	if (status == NSS_STATUS_TRYAGAIN && *errnop == ERANGE)
	    return status;

	if (status == NSS_STATUS_SUCCESS)
	    return NSS_STATUS_SUCCESS;
    } while (__nss_next(&grent_nip, grent_fct_name, &grent_fct.ptr, status, 0) == 0);

    grent_nip = NULL;
    return NSS_STATUS_NOTFOUND;
}


enum nss_status
_nss_nonlocal_getgrnam_r(const char *name, struct group *grp,
			 char *buffer, size_t buflen, int *errnop)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = &__nss_group_nonlocal_lookup, .fct_name = "getgrnam_r",
	.status = &status, .errnop = errnop
    };
    const __typeof__(&_nss_nonlocal_getgrnam_r) self = NULL;

    char *nonlocal_ignore = getenv(NONLOCAL_IGNORE_ENV);
    if (nonlocal_ignore != NULL && nonlocal_ignore[0] != '\0')
	return NSS_STATUS_UNAVAIL;

#define args (name, grp, buffer, buflen, errnop)
#include "walk_nss.h"
#undef args
    if (status != NSS_STATUS_SUCCESS)
	return status;

    if (strcmp(name, grp->gr_name) != 0) {
	syslog(LOG_ERR, "nss_nonlocal: discarding group %s from lookup for group %s\n", grp->gr_name, name);
	return NSS_STATUS_NOTFOUND;
    }

    return check_nonlocal_group(name, grp, errnop);
}

enum nss_status
_nss_nonlocal_getgrgid_r(gid_t gid, struct group *grp,
			 char *buffer, size_t buflen, int *errnop)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = &__nss_group_nonlocal_lookup, .fct_name = "getgrgid_r",
	.status = &status, .errnop = errnop
    };
    const __typeof__(&_nss_nonlocal_getgrgid_r) self = NULL;

    char *nonlocal_ignore = getenv(NONLOCAL_IGNORE_ENV);
    if (nonlocal_ignore != NULL && nonlocal_ignore[0] != '\0')
	return NSS_STATUS_UNAVAIL;

#define args (gid, grp, buffer, buflen, errnop)
#include "walk_nss.h"
#undef args
    if (status != NSS_STATUS_SUCCESS)
	return status;

    if (gid != grp->gr_gid) {
	syslog(LOG_ERR, "nss_nonlocal: discarding gid %d from lookup for gid %d\n", grp->gr_gid, gid);
	return NSS_STATUS_NOTFOUND;
    }

    return check_nonlocal_group(grp->gr_name, grp, errnop);
}

static bool
add_group(gid_t group, long int *start, long int *size, gid_t **groupsp,
	  long int limit, int *errnop, enum nss_status *status)
{
    int i, old_errno = errno;
    for (i = 0; i < *start; ++i)
	if ((*groupsp)[i] == group)
	    return true;
    if (*start + 1 > *size) {
	gid_t *newgroups;
	long int newsize = 2 * *size;
	if (limit > 0) {
	    if (*size >= limit) {
		*status = NSS_STATUS_SUCCESS;
		return false;
	    }
	    if (newsize > limit)
		newsize = limit;
	}
	newgroups = realloc(*groupsp, newsize * sizeof((*groupsp)[0]));
	errno = old_errno;
	if (newgroups == NULL) {
	    *errnop = ENOMEM;
	    *status = NSS_STATUS_TRYAGAIN;
	    return false;
	}
	*groupsp = newgroups;
	*size = newsize;
    }
    (*groupsp)[(*start)++] = group;
    return true;
}

enum nss_status
_nss_nonlocal_initgroups_dyn(const char *user, gid_t group, long int *start,
			     long int *size, gid_t **groupsp, long int limit,
			     int *errnop)
{
    enum nss_status status;
    const struct walk_nss w = {
	.lookup = &__nss_group_nonlocal_lookup, .fct_name = "initgroups_dyn",
	.status = &status, .errnop = errnop
    };
    const __typeof__(&_nss_nonlocal_initgroups_dyn) self = NULL;

    struct group local_users_group, nonlocal_users_group;
    bool is_nonlocal = true;
    char *buffer;
    int in, out, i;

    /* Check that the user is a nonlocal user, or a member of the
     * MAGIC_NONLOCAL_GROUPNAME group, before adding any groups. */
    status = check_nonlocal_user(user, errnop);
    if (status == NSS_STATUS_TRYAGAIN) {
	return status;
    } else if (status != NSS_STATUS_SUCCESS) {
	is_nonlocal = false;

	status = get_local_group(MAGIC_LOCAL_GROUPNAME,
				 &local_users_group, &buffer, errnop);
	if (status == NSS_STATUS_SUCCESS) {
	    free(buffer);
	    if (!add_group(local_users_group.gr_gid, start, size, groupsp,
			   limit, errnop, &status))
		return status;
	} else if (status == NSS_STATUS_TRYAGAIN) {
	    return status;
	} else {
	    syslog(LOG_WARNING,
		   "nss_nonlocal: Group %s does not exist locally!",
		   MAGIC_LOCAL_GROUPNAME);
	}
    }

    status = get_local_group(MAGIC_NONLOCAL_GROUPNAME,
			     &nonlocal_users_group, &buffer, errnop);
    if (status == NSS_STATUS_SUCCESS) {
	free(buffer);
	if (is_nonlocal) {
	    if (!add_group(nonlocal_users_group.gr_gid, start, size, groupsp,
			   limit, errnop, &status))
		return status;
	} else {
	    int i;
	    for (i = 0; i < *start; ++i) {
		if ((*groupsp)[i] == nonlocal_users_group.gr_gid) {
		    is_nonlocal = true;
		    break;
		}
	    }

	    if (is_nonlocal) {
		struct passwd pwbuf;
		char *buf;
		int nonlocal_errno = *errnop;
		status = get_nonlocal_passwd(user, &pwbuf, &buf, errnop);

		if (status == NSS_STATUS_SUCCESS) {
		    nonlocal_errno = *errnop;
		    status = check_nonlocal_gid(user, NULL, pwbuf.pw_gid,
						&nonlocal_errno);
		    free(buf);
		}

		if (status == NSS_STATUS_SUCCESS) {
		    if (!add_group(pwbuf.pw_gid, start, size, groupsp, limit,
				   errnop, &status))
			return status;
		} else if (status == NSS_STATUS_TRYAGAIN) {
		    *errnop = nonlocal_errno;
		    return status;
		}
	    }
	}
    } else if (status == NSS_STATUS_TRYAGAIN) {
	if (is_nonlocal)
	    return status;
    } else {
	syslog(LOG_WARNING, "nss_nonlocal: Group %s does not exist locally!",
	       MAGIC_NONLOCAL_GROUPNAME);
    }

    if (!is_nonlocal)
	return NSS_STATUS_SUCCESS;

    in = out = *start;

#define args (user, group, start, size, groupsp, limit, errnop)
#include "walk_nss.h"
#undef args
    if (status != NSS_STATUS_SUCCESS)
        return status;

    for (; in < *start; ++in) {
	int nonlocal_errno = *errnop;

	for (i = 0; i < out; ++i)
	    if ((*groupsp)[i] == (*groupsp)[in])
		break;
	if (i < out)
	    continue;

	status = check_nonlocal_gid(user, NULL, (*groupsp)[in],
				    &nonlocal_errno);
	if (status == NSS_STATUS_SUCCESS) {
	    (*groupsp)[out++] = (*groupsp)[in];
	} else if (status == NSS_STATUS_TRYAGAIN) {
	    *start = out;
	    *errnop = nonlocal_errno;
	    return status;
	}
    }

    *start = out;
    return NSS_STATUS_SUCCESS;
}
